### PR TITLE
Better filter UI

### PR DESCRIPTION
**Pull request by bvaughn**
Builds on top of PR #640 and...
* Adds inline search highlights so it's obvious at a glance that you're viewing a _filtered_ set of the data and not the full tree.
* Tidies up alignment of toolbar
* Makes focused state of search input look more like Chrome search and less aqua
* Make the cancel button a little more subtle and improve its vertical alignment
* Adds search icon to search input
* Removes regex checkbox in favor of auto-detection based on search text (eg "foo" vs "/foo")
* Added error handling for invalid RegExps

## Input states

Default | Focused | Invalid regex
--- | --- | ---
screen shot 2017-04-27 at 7 22 42 pm | screen shot 2017-04-27 at 7 22 59 pm | screen shot 2017-04-27 at 7 23 35 pm

## Plaintext search
plaintext search demo

## RegExp search
RegExp search demo

### Resizing
resize demo

#### Chrome
Chrome screenshot

#### Safari
Safari screenshot

#### Firefox
Firefox screenshot

#### Electron
Electron screenshot

### CODE REVIEW

**Comment by gaearon** [comment]
Do you think it's worth making the input font size smaller and tweaking the clear button? Now that there are other indicators that search is active.

**Comment by bvaughn** [comment]
Yup! I actually already shrunk the font size a little. I was fighting the urge to shrink it more because I know I have a tendency to make things too small. But let me try making it a little smaller. :grin:

**Comment by bvaughn** [comment]
Updated demo video.

**Comment by gaearon** [comment]
Wanna add some border radius to the input?

**Comment by gaearon** [comment]
Also what do you think about moving input + search-related checkboxes to the right?

**Comment by gaearon** [comment]
And maybe adding a magnifying glass watermark to the search icon a la

screen shot 2017-04-27 at 21 27 08

**Comment by bvaughn** [comment]
> Also what do you think about moving input + search-related checkboxes to the right?

Thought about this! Seems great when they're all on the same line but a bit weird once they wrap (as they're doing in my video). Might still be worth doing though. Let me try it.

**Comment by gaearon** [comment]
We can lay out them like now if they're small enough, but do a more sensible layout on larger screens.

**Comment by bvaughn** [comment]
How about this?
resize-search-options

**Comment by gaearon** [comment]
A few things still look off (not sure if fixable with small changes or not):

* I'm not sure if Highlight Search is more related to search or not.
* In general it is annoying that a niche feature like search by regex is taking so much space.

I wonder if these two search options would be better as toggleable buttons with icons, like in some IDEs. Not sure if it's going too far though. But I'm not very happy with how we spend this entire top bar on checkboxes people click very rarely.

**Comment by bvaughn** [comment]
Yeah, I don't know. Highlight search _is_ related to the search but... to it's results, more-so than the criteria... so I thought to leave it on its own.

Agreed about space. How about just "RegExp" instead of "Use Regular Expressions"?

What about this:
search-input

**Comment by iamdustan** [comment]
What about using a command like prefixing with `/` to switch to RegExp mode?

* No UI needed.
* Reasonable default experience.
* Power-user feature.
* Low-cost.

**Comment by bvaughn** [comment]
Hm, initial reaction: That isn't very discoverable. (It's not how browser search or IDE search works.) And "[ ] RegExp" is pretty small now so it doesn't take up much space.

**Comment by gaearon** [comment]
Could use a placeholder for this.

**Comment by bvaughn** [comment]
Now the search input takes up way more space than the checkbox + input currently does though

**Comment by bvaughn** [comment]
Let me whip it up quickly and see what it looks like though :smile:

**Comment by gaearon** [comment]
We could also omit "components", it's pretty obvious what it searches for now that we have highlighting.

screen shot 2017-04-27 at 22 17 08

**Comment by bvaughn** [comment]
I'm sold 😁 

regex-in-placeholder

**Comment by gaearon** [comment]
If you feel like bikeshedding a little bit more, another random idea:

screen shot 2017-04-27 at 22 23 33

**Comment by gaearon** [comment]
@bvaughn Anyway, I think your last version is best. Let's just rename "Trace React Updates" to "Highlight Updates".

**Comment by bvaughn** [comment]
Thanks for the feedback folks!

type-ahead-demo-final

File: frontend/SearchPane.js [hunk 1]
```diff
@@ -36,155 +33,27 @@ class SearchPane extends React.Component {
   _key: (evt: EventLike) => void;
```
Comment (gaearon) [inline frontend/SearchPane.js]: Should probably move this Flow annotation (and maybe rename this file?)
I didn't really completely clean this code up.

**Comment by gaearon** [comment]
While not quite related, can we also debounce search input just a little bit?

**Comment by gaearon** [comment]
It seems like `/` matches all components now. Would be nice to avoid this as I imagine it could be very slow to type.

**Comment by bvaughn** [comment]
Good call. I'll add a short-circuit of the search text is equal to "/" 😁 

I find myself wanting to do a bit of cleanup here but I'd like to limit the footprint of this specific PR. Happy to add the debouncing change and the rename you mentioned above. Maybe we could cap it there and do follow up with another PR?

**Comment by gaearon** [comment]
Agreed, but let's fix `/` here.

**Comment by bvaughn** [comment]
Certainly!

File: frontend/SearchPane.js (hunk 1 above)
Comment (bvaughn) [inline frontend/SearchPane.js]: Good call.

**Comment by bvaughn** [comment]
Okay. Tidied up the old `SearchPane` and added better edge-case handling around "/" and "//" strings.

Maybe we can leave off the debounce for now. I think I might like to revisit how our search index and results ranking works in a subsequent PR anyway.

**Comment by gaearon** [comment]
When I click "Highlight Search" the placeholder makes a weird change.
Can we completely remove this behavior? (I assume it was added together with the checkbox.)

screen shot 2017-04-27 at 23 56 06

**Comment by bvaughn** [comment]
Oh. Yeah sure.

I only tested that feature briefly and I didn't notice the placeholder change.

**Comment by gaearon** [comment]
A few other things (for follow up):

* Perhaps `/` along should be treated as an empty search input (no filtering) rather than "no results".
* I think we should change the filtering to only work at word boundaries. For example `to` should find `TodoItem` and `AddTodo`, but not `button`.

**Comment by bvaughn** [comment]
> Perhaps / along should be treated as an empty search input (no filtering) rather than "no results".

Sure.

> I think we should change the filtering to only work at word boundaries. For example to should find TodoItem and AddTodo, but not button.

I'd rather do this as a follow-up.

**Comment by gaearon** [comment]
Yea, totally, separately sounds good.

**Comment by bvaughn** [comment]
slash-handling

**Comment by gaearon** [comment]
Hmm, it doesn't let me fully type a regex. Try typing `/t[`. At this point it stays `/t`.

**Comment by bvaughn** [comment]
Not at my keyboard at the moment but I bet I know the cause of this. Previous UI didn't try to convert the search term into a regex until you clicked the checkbox. Now we convert as you type if there's a forward slash at the start. I bet I need to try catch the conversion for temporarily-invalid regular expressions (like the one you've typed above). Maybe also add a red background color to the search input box to convey that the regular expression is incorrect. I'll take another look at it this evening. Nice catch exclamation!

**Comment by bvaughn** [comment]
lol at the speech to text for writing out the word exclamation

**Comment by bvaughn** [comment]
Hm...actually what you've described ("/" => "/t" => "/t[") works for me.

**Comment by gaearon** [comment]
Weird, it doesn't for me. I'm testing in the plain shell on the default example (`shells/plain`). It seems like I can't type `[` at all if it's in regex mode.

**Comment by bvaughn** [comment]
Curious if you can repro in master (after the "regex" checkbox is selected). I assume you would be able to.

**Comment by bvaughn** [comment]
How about this for invalid regexps?

invalid-regexp

**Comment by gaearon** [comment]
This works really well. The last issue I can spot is search icon in Firefox:

screen shot 2017-04-28 at 10 15 24

I’ll take a look at it.

**Comment by bvaughn** [comment]
Weird! Did you pull before checking out Firefox? I tested it (and Safari) and left screenshots in the PR description yesterday.

**Comment by gaearon** [comment]
Yea, I definitely pulled. Maybe a different version or something. 🤷‍♀️

**Comment by gaearon** [comment]
It was just missing a `top: 0`.

**Comment by bvaughn** [comment]
Weird! 🤷‍♂️ Worked for my Safari.

Glad you caught it!

**Pull request merged**